blas/gonum/level2cmplx128.go

```diff
@@ -1239,6 +1239,201 @@ func (Implementation) Zhpr2(uplo blas.Uplo, n int, alpha complex128, x []complex
 	}
 }
 
+// Ztbmv performs one of the matrix-vector operations
+//  x = A * x    if trans = blas.NoTrans
+//  x = A^T * x  if trans = blas.Trans
+//  x = A^H * x  if trans = blas.ConjTrans
+// where x is an n element vector and A is an n×n triangular band matrix, with
+// (k+1) diagonals.
+func (Implementation) Ztbmv(uplo blas.Uplo, trans blas.Transpose, diag blas.Diag, n, k int, ab []complex128, ldab int, x []complex128, incX int) {
+	if uplo != blas.Upper && uplo != blas.Lower {
+		panic(badUplo)
+	}
+	if trans != blas.NoTrans && trans != blas.Trans && trans != blas.ConjTrans {
+		panic(badTranspose)
+	}
+	if diag != blas.Unit && diag != blas.NonUnit {
+		panic(badDiag)
+	}
+	checkZtbMatrix('A', n, k, ab, ldab)
+	checkZVector('x', n, x, incX)
+
+	if n == 0 {
+		return
+	}
+
+	// Set up start index in X.
+	var kx int
+	if incX < 0 {
+		kx = (1 - n) * incX
+	}
+
+	switch trans {
+	case blas.NoTrans:
+		if uplo == blas.Upper {
+			if incX == 1 {
+				for i := 0; i < n; i++ {
+					kk := min(k, n-i-1)
+					if diag == blas.NonUnit {
+						x[i] *= ab[i*ldab]
+					}
+					for j, aij := range ab[i*ldab+1 : i*ldab+kk+1] {
+						x[i] += x[i+j+1] * aij
+					}
+				}
+			} else {
+				ix := kx
+				for i := 0; i < n; i++ {
+					kk := min(k, n-i-1)
+					if diag == blas.NonUnit {
+						x[ix] *= ab[i*ldab]
+					}
+					jx := ix + incX
+					for _, aij := range ab[i*ldab+1 : i*ldab+kk+1] {
+						x[ix] += x[jx] * aij
+						jx += incX
+					}
+					ix += incX
+				}
+			}
+		} else {
+			if incX == 1 {
+				for i := n - 1; i >= 0; i-- {
+					kk := min(k, i)
+					if diag == blas.NonUnit {
+						x[i] *= ab[i*ldab+k]
```

Comment:
Is there a benefit to working on `tmp := x[i]` and restoring in these loops? This takes potentially `n` bounds checks to 2 bounds checks for these loops.

Reply:
There probably is although I didn't measure it. Originally I wrote it for conciseness but also with the temporary variables it's clear enough, so I've changed the code as you suggest.